models/attachment.go

```diff
@@ -1,11 +1,151 @@
+
+	// If we've already determined there are no template variables in this attachment return it immediately
+	if a.vanillaFile == true {
+		return decodedAttachment, nil
+	} else {
+
+		// Decided to use the file extension rather than the content type, as there seems to be quite
+		//  a bit of variability with types. e.g sometimes a Word docx file would have:
+		//   "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
+		fileExtension := filepath.Ext(a.Name)
+
+		switch fileExtension {
+
+		case ".docx", ".docm", ".pptx", ".xlsx", ".xlsm":
+			// Most modern office formats are xml based and can be unarchived.
+			// .docm and .xlsm files are comprised of xml, and a binary blob for the macro code
+
+			// Zip archives require random access for reading, so it's hard to stream bytes. Solution seems to be to use a buffer.
+			// See https://stackoverflow.com/questions/16946978/how-to-unzip-io-readcloser
+			b := new(bytes.Buffer)
+			b.ReadFrom(decodedAttachment)
+			buf := b.Bytes()
+			zipReader, err := zip.NewReader(bytes.NewReader(buf), int64(len(buf))) // Create a new zip reader from the file
+			if err != nil {
+				return nil, err
+			}
+
+			newZipArchive := new(bytes.Buffer)
+			zipWriter := zip.NewWriter(newZipArchive) // For writing the new archive
+
+			// i. Read each file from the Word document archive
+			// ii. Apply the template to it
+			// iii. Add the templated content to a new zip Word archive
+			fileContainedTemplatesVars := false
+			for _, zipFile := range zipReader.File {
+				ff, err := zipFile.Open()
+				if err != nil {
+					return nil, err
+				}
+				defer ff.Close()
+				contents, err := ioutil.ReadAll(ff)
+				if err != nil {
+					return nil, err
+				}
+				subFileExtension := filepath.Ext(zipFile.Name)
+				var tFile string
+				if subFileExtension == ".xml" || subFileExtension == ".rels" { // Ignore other files, e.g binary ones and images
+					// For each file apply the template.
+					tFile, err = ExecuteTemplate(string(contents), ptx)
+					if err != nil {
+						return nil, err
+					}
+					// Check if the subfile changed. We only need this to be set once to know in the future to check the 'parent' file
+					if tFile != string(contents) {
+						fileContainedTemplatesVars = true
+					}
+				} else {
+					tFile = string(contents) // Could move this to the declaration of tFile, but might be confusing to read
+				}
+				// Write new Word archive
+				newZipFile, err := zipWriter.Create(zipFile.Name)
+				if err != nil {
+					zipWriter.Close() // Don't use defer when writing files https://www.joeshaw.org/dont-defer-close-on-writable-files/
+					return nil, err
+				}
+				_, err = newZipFile.Write([]byte(tFile))
+				if err != nil {
+					zipWriter.Close()
+					return nil, err
+				}
+			}
+			// If no files in the archive had template variables, we set the 'parent' file to not be checked in the future
+			if fileContainedTemplatesVars == false {
```

Comment:
I'm still hoping that I can come up with a simpler option (e.g. looking at `sync.Once` as an alternative), but in the meantime, I think we can assign `a.vannillaFile` above and not use `fileContainedTemplatesVars`?

Reply:
I can't remember why I used `fileContainedTemplatesVars`. Removed.